### DISCUSSION TITLE
Feature request: detailed discussion on achieving good performance building proofs

### DISCUSSION THREAD

**Discussion by robertsdotpm** [risc0/risc0#1887] [General]
Hello people, I really love this project, it seems to me to be literal witch craft and the number of new cryptographic protocols this could facilitate is astonishing. What seems to me to be missing from this project is a page in the documentation that discusses what hardware aspects impact the speed of proofs and how to achieve better performance with proof generation. For example -- does the VM benefit from really high clock speed? Does it benefit from more cores? Is the VM concurrent? Do the majority of the examples utilize concurrency in the VM? Is CUDA even worthwhile to setup? I really don't know but any details about performance would really go a long way in encouraging more people to use this project.

As a first-time user it seems that when you build the examples a huge amount of time is spent actually building the VM. My assumption was that this amount of time would be required every time you ran the proof but this was wrong -- proof time dropped significantly once the artefacts were built. Mentioning this in the docs would be a good idea. I wonder also if the actual build artefacts could be cached for this so new users didn't have to compile the whole thing each time.

The other thing I wanted to say is compared to the benchmark page my performance was significantly slower (47s vs 2s for json on the m2.) Granted, I did just use a CPU, and I think the benchmark numbers are mostly for GPUs. Again -- it would be worthwhile writing a note some where that says something like 'these are all for GPUs / graphics cards. Running proofs with a standard CPU will lead to significantly slower performance.' Or something like that. We definitely need more information on achieving the best performance for proof generation, IMO.

**Comment by nuke-web3** [comment]
Of note: https://dev.risczero.com/api/zkvm/profiling and the following pages discuss strategies to evaluate and improve perf, but could be improved for sure.
